### PR TITLE
Move from upstream go to microsoft go

### PR DESCRIPTION
**Pull request by Kaderinho**
The main motivation is to use Microsoft Go to enable FIPS mode in the Agent.

To do so, we download both toolchains and add code into the bashrc to select the right toolchain at runtime (defaulting to Microsoft Go)
~Also, there is a PR to add a [nightly pipeline to create buildimages](https://github.com/DataDog/gitlab-config/pull/165/files) that uses Go Upstream. Note since both toolchain are installed, GOROOT can be changed so we technically can compile using Go upstream or Microsoft Go with a same buildimage.~

### CODE REVIEW

File: system-probe_arm64/Dockerfile [hunk 1]
```diff
@@ -56,12 +58,19 @@ ENV PATH "${CONDA_PATH}/condabin:${PATH}"
...
+    && wget -O /tmp/golang.tar.gz https://go.dev/dl/go$GO_VERSION.linux-arm64.tar.gz \
+    && echo "$GO_SHA256_LINUX_ARM64  /tmp/golang.tar.gz" | sha256sum --check \
+    && tar -C /usr/local -xzf /tmp/golang.tar.gz \
+    && rm -f /tmp/golang.tar.gz \
+    && echo "Installing Microsoft Go" \
+    && wget -O /tmp/golang.tar.gz https://aka.ms/golang/release/latest/go${GO_VERSION}.linux-arm64.tar.gz \
+    && echo "$MSGO_SHA256_LINUX_ARM64  /tmp/golang.tar.gz" | sha256sum --check \
+    && mkdir /usr/local/msgo && tar --strip-components=1 -C /usr/local/msgo/ -xzf /tmp/golang.tar.gz \
+    && rm -f /tmp/golang.tar.gz
+ENV PATH "${GOROOT}/bin:${PATH}"
```
Comment (brycekahle) [inline system-probe_arm64/Dockerfile:71]: This line duplicates the line below

Comment (brycekahle) [inline system-probe_arm64/Dockerfile:71]: Can you combine this with the line below?

File: system-probe_x64/Dockerfile [hunk 2]
```diff
@@ -61,12 +63,19 @@ ENV PATH "${CONDA_PATH}/condabin:${PATH}"
...
+    && wget -O /tmp/golang.tar.gz https://go.dev/dl/go$GO_VERSION.linux-amd64.tar.gz \
+    && echo "$GO_SHA256_LINUX_AMD64  /tmp/golang.tar.gz" | sha256sum --check \
+    && tar -C /usr/local -xzf /tmp/golang.tar.gz \
+    && rm -f /tmp/golang.tar.gz \
+    && echo "Installing Microsoft Go" \
+    && wget -O /tmp/golang.tar.gz https://aka.ms/golang/release/latest/go${GO_VERSION}.linux-amd64.tar.gz \
+    && echo "$MSGO_SHA256_LINUX_AMD64  /tmp/golang.tar.gz" | sha256sum --check \
+    && mkdir /usr/local/msgo && tar --strip-components=1 -C /usr/local/msgo/ -xzf /tmp/golang.tar.gz \
+    && rm -f /tmp/golang.tar.gz
+ENV PATH "${GOROOT}/bin:${PATH}"
```
Comment (brycekahle) [inline system-probe_x64/Dockerfile:76]: Can you combine this with the line below?

File: go.env [hunk 3]
```diff
@@ -1,5 +1,9 @@
+MSGO_SHA256_LINUX_AMD64=d49fda9363d5a4a0d0d7060447c25b197ebe57574d80892efef2e2768705a59d
+MSGO_SHA256_LINUX_ARM64=2e7164a1be4b90982568ac5bf94050a3384271bc0341c08783a668e74073f2fb
+MSGO_SHA256_LINUX_ARMV6L=8f622cf1062095db07fb224471c75c8a8c8e7c138f29a53fd838a2a105bbbf39
+MSGO_SHA256_WINDOWS_AMD64=1067c9490e8914f783165665c8dab10f11018d3cc3df7f6e0baa10621e654626
 GO_VERSION=1.22.6
 GO_SHA256_LINUX_AMD64=999805bed7d9039ec3da1a53bfbcafc13e367da52aa823cb60b68ba22d44c616
 GO_SHA256_LINUX_ARM64=c15fa895341b8eaf7f219fada25c36a610eb042985dc1a912410c1c90098eaf2
 GO_SHA256_LINUX_ARMV6L=b566484fe89a54c525dd1a4cbfec903c1f6e8f0b7b3dbaf94c79bc9145391083
-GO_SHA256_WINDOWS_AMD64=6023083a6e4d3199b44c37e9ba7b25d9674da20fd846a35ee5f9589d81c21a6a
+GO_SHA256_WINDOWS_AMD64=6023083a6e4d3199b44c37e9ba7b25d9674da20fd846a35ee5f9589d81c21a6a
```
Comment (brycekahle) [inline go.env:9]: nit: the trailing newline was removed

File: rpm-armhf/Dockerfile
Comment (chouquette) [inline rpm-armhf/Dockerfile:1]: Not strictly related to your PR, but I'm surprised we use an armv6 build for `armhf`
armhf is supposed to be armv7 so we might use a more efficient toolchain here

**Review by pgimalac — APPROVED** [review]
LGTM

File: tasks/update_go.py [hunk 4]
```diff
@@ -63,6 +63,34 @@ def _check_archive(version: str, shas: List[Tuple[Platform, str]]):
...
+
+    assets = _fetch_msgo_assets(version)
+    shas = _get_msgo_sha256(version, assets)
+    return shas
+
+
+def _display_shas(shas: List[Tuple[Platform, str]], toolchain: str):
+    print(f"--- {toolchain} ---")
+    for (os, arch), sha in shas:
+        print(f"[{os}/{arch}] {sha}")
```
[the remark this answers is not in the document]
Comment (Kaderinho) [reply tasks/update_go.py:93]: hmmmm I don't remember when I deleted this but I definitely wanted to keep the alignment mb 😬 (I swear I did not want to erase you great work :p)

**Review by brycekahle — APPROVED** [review]

File: windows/Dockerfile [hunk 5]
```diff
@@ -23,6 +23,8 @@ ENV WINDOWS_VERSION=${WINDOWS_VERSION:-1809}
 ARG DD_TARGET_ARCH
 ENV TARGET_ARCH=${DD_TARGET_ARCH:-x64}
 
+ARG GOROOT
+
```
Comment (chouquette) [inline windows/Dockerfile:27]: Is this a WIP leftover, or is this still needed for Windows?

Comment (Kaderinho) [reply windows/Dockerfile:27]: Still needed for now, value is populated from `build-container.ps1`. I'll modify that to use the same runtime logic as the linux images but for now I want to unblock the datadog-agent PR

**Comment by Kaderinho** [comment]
Close in favor of https://github.com/DataDog/datadog-agent-buildimages/tree/nicolas.guerguadj/add-msgo-to-buildimages

**Pull request closed without merging**